### ISSUE TITLE
No need for SpaceConsistencyBear to run if PEP8Bear is also passed into the bears argument

### ISSUE THREAD

**Issue by Naveenaidu** [coala/coala#5241]
PEP8Bear also takes into consideration of `Trailing Whitespaces` just as `SpaceConsistencyBear` does. So it automatically fixes the patch. And because of that when `SpaceConsistencyBear` starts running from the process queue it does show that there are errors, but shows no option for the user to `Apply a Patch` since that patch was already applied by the PEP8Bear. This might create confusion among the users.

And IMO fixing this might also increase the performance of coala .

The Sample outputs  when both `SpaceConsistencyBear` and `PEP8Bear` are used:-
```
test.py
[   3] def·test(a1,b1="",c1="",a2):·
**** PEP8Bear [Section: cli | Severity: NORMAL] ****
!    ! The code does not comply to PEP8.
[----] /home/theprophet/test-coala/test.py
[++++] /home/theprophet/test-coala/test.py
[   3] def test(a1,b1="",c1="",a2): 
[   3] def test(a1, b1="", c1="", a2):
[    ] *0. Do (N)othing
[    ]  1. (O)pen file
[    ]  2. (A)pply patch
[    ]  3. Add (I)gnore comment
[    ]  4. Show Applied (P)atches
[    ]  5. (G)enerate patches
[    ] Enter number (Ctrl-D to exit): 2
[    ] Patch applied successfully.
[    ] *0. Do (N)othing
[    ]  1. (O)pen file
[    ]  2. Add (I)gnore comment
[    ] Enter number (Ctrl-D to exit): 0

test.py
[   3] def·test(a1,b1="",c1="",a2):·
**** SpaceConsistencyBear [Section: cli | Severity: NORMAL] ****
!    ! Line contains following spacing inconsistencies:
!    ! - Trailing whitespaces.
[    ] *0. Do (N)othing
[    ]  1. (O)pen file
[    ]  2. Print (M)ore info
[    ]  3. Add (I)gnore comment
[    ] Enter number (Ctrl-D to exit): 0
```
The `Trailing whitespaces` was already fixed by the PEP8Bear, but this issue comes up again in `SpaceConsistencyBear` again but now with no options to apply,since it was already fixed. IMO there is no need for the `SpaceConsistencyBear` to run here. It simply slows down the coala and creates confusion.

The output when the file runs only with `SpaceConsistencyBear`
```
test.py
[   3] def·test(a1,·b1="",·c1="",·a2):·
**** SpaceConsistencyBear [Section: cli | Severity: NORMAL] ****
!    ! Line contains following spacing inconsistencies:
!    ! - Trailing whitespaces.
[----] /home/theprophet/test-coala/test.py
[++++] /home/theprophet/test-coala/test.py
[   3] def test(a1, b1="", c1="", a2): 
[   3] def test(a1, b1="", c1="", a2):
[    ] *0. Do (N)othing
[    ]  1. (O)pen file
[    ]  2. (A)pply patch
[    ]  3. Print (M)ore info
[    ]  4. Add (I)gnore comment
[    ]  5. Show Applied (P)atches
[    ]  6. (G)enerate patches
[    ] Enter number (Ctrl-D to exit): 2
[    ] Patch applied successfully.
[    ] *0. Do (N)othing
[    ]  1. (O)pen file
[    ]  2. Print (M)ore info
[    ]  3. Add (I)gnore comment
[    ] Enter number (Ctrl-D to exit): 0

```

status\needs discussion
@Makman2  Need your opinion.

**Comment by Naveenaidu** [comment]
Please Correct me if i am wrong,The bears get the original file ,not the file where the changes are being made? ,So When the process queue started `PEP8Bear` it  made changes/applied patches to the diff file/virtual file stored on the memory rather than the original file. And when it was the turn of `SpaceConsistencyBear` to start, It was passed the original file and since the `Trailing Whitespace` is still present in the original file, it showed a error on that file. 

The reason for `Apply Patch` not showing up as option for `SpaceConsistencyBear` is because these actions are controlled by the `ConsoleInteraction.py` and `Apply Patch` only shows up if there was no actions ever applied on that line of the file(Because of the original diff handling method). Is this correct?

If that is correct, how about we add a if/else condition when the bears are put into the processing que and not pass `SpaceConsistencyBear` if PEP8Bear is present.

**Comment by Makman2** [comment]
> The bears get the original file ,not the file where the changes are being made? 

Yes. Inside a single section all bears operate in parallel on the original file data.

> The reason for Apply Patch not showing up as option for SpaceConsistencyBear is because these actions are controlled by the ConsoleInteraction.py and Apply Patch only shows up if there was no actions ever applied on that line of the file(Because of the original diff handling method). Is this correct?

I think not quite, I believe it's due to `is_applicable()` inside the `ApplyPatchAction`. It sees there's already a patch applied, and the same patch from the `SpaceConsistencyBear` produces a conflict then, so the option isn't offered.

> If that is correct, how about we add a if/else condition when the bears are put into the processing que and not pass SpaceConsistencyBear if PEP8Bear is present.

That's a quite bigger problem. Aspects shall solve that, however they are far from being ready. A specific switch for the `SpaceConsistencyBear` is not okay though, this has to work generically for other bears too in case this is needed.

I would pledge to have this as wontfix, you can "cleverly" combine bears in different sections and disable/enable settings on bears to achieve what you want. This is not elegant, but only aspects are a future-proof solution to avoid this possible duplication.

Your opinions @Naveenaidu ?

**Comment by Naveenaidu** [comment]
But why do you think adding a if/else condition is a big problem. Sorry, if this question is stupid, i am new to coala and it'll still take me some time to understand it's working properly  :\

**Comment by Makman2** [comment]
> test.py
[   3] def·test(a1,·b1="",·c1="",·a2):·
**** SpaceConsistencyBear [Section: cli | Severity: NORMAL] ****
!    ! SpaceConsistencyBear and PEP8Bear have same outputs

This is not possible. Bears work independently from each other.

> But why do you think adding a if/else condition is a big problem. Sorry, if this question is stupid, i am new to coala and it'll still take me some time to understand it's working properly :\

You would have to hardcode that for each bear combination.
1. Impossible to maintain 
2. Highly inconsistent, as you can write custom bears. How to make things consistent for them, in case they have also same analysis running like other bears?
3. coala shouldn't be aware of specific bear implementations, it just offers a framework.

**Comment by Naveenaidu** [comment]
>- This is not possible. Bears work independently from each other.
Sorry, It completely slipped of my mind. That was a stupid question to ask though 😅 .
Now, I understand the reasons. I think it's a wont fix then. Please label the issue as `won't fix`.
